pkg/controllers/infracluster/infracluster_controller.go

```diff
@@ -0,0 +1,189 @@
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strconv"
+
+	"github.com/go-logr/logr"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+
+	"k8s.io/client-go/rest"
+	awsv1 "sigs.k8s.io/cluster-api-provider-aws/v2/api/v1beta2"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/cluster-capi-operator/e2e/framework"
+	"github.com/openshift/cluster-capi-operator/pkg/controllers"
+	"github.com/openshift/cluster-capi-operator/pkg/operatorstatus"
+)
+
+const (
+	// Controller conditions for the Cluster Operator resource
+	InfraClusterControllerAvailableCondition = "InfraClusterControllerAvailable"
+	InfraClusterControllerDegradedCondition  = "InfraClusterControllerDegraded"
+
+	defaultCAPINamespace              = "openshift-cluster-api"
+	providerConfigMapLabelVersionKey  = "provider.cluster.x-k8s.io/version"
+	providerConfigMapLabelTypeKey     = "provider.cluster.x-k8s.io/type"
+	providerConfigMapLabelNameKey     = "provider.cluster.x-k8s.io/name"
+	ownedProviderComponentName        = "cluster.x-k8s.io/provider"
+	imagePlaceholder                  = "to.be/replaced:v99"
+	openshiftInfrastructureObjectName = "cluster"
+	defaultInfraClusterName           = "cluster"
+	notNamespaced                     = ""
+	clusterOperatorName               = "cluster-api"
+	defaultCoreProviderComponentName  = "cluster-api"
+)
+
+type InfraClusterController struct {
+	operatorstatus.ClusterOperatorStatusClient
+	Scheme   *runtime.Scheme
+	Images   map[string]string
+	RestCfg  *rest.Config
+	Platform configv1.PlatformType
+	Infra    configv1.Infrastructure
+}
+
+// Reconcile reconciles the cluster-api ClusterOperator object.
+func (r *InfraClusterController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx).WithName("InfraClusterController")
+
+	res, err := r.reconcile(ctx, log)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("error during reconcile: %w", err)
+	}
+
+	if err := r.setAvailableCondition(ctx, log); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to set conditions for InfraCluster Controller: %w", err)
+	}
+
+	return res, nil
+}
+
+func (r *InfraClusterController) createAWSCluster(ctx context.Context) error {
+
+	apiUrl, err := url.Parse(r.Infra.Status.APIServerInternalURL)
+	if err != nil {
+		return fmt.Errorf("failed to parse apiUrl: %w", err)
+	}
+
+	port, err := strconv.ParseInt(apiUrl.Port(), 10, 32)
+	if err != nil {
+		return fmt.Errorf("failed to parse apiUrl port: %w", err)
+	}
+
+	if r.Infra.Status.PlatformStatus == nil {
+		return fmt.Errorf("infrastructure PlatformStatus should not be nil: %w", err)
+	}
+
+	awsCluster := &awsv1.AWSCluster{
+		ObjectMeta: metav1.ObjectMeta{
```

Comment:
Do we need to annotate this so that it is not managed by the infracluster controller?

Reply:
This PR was still WIP back then. This comment was addressed in the implementation since then.